### PR TITLE
feat: compile training loop automatically using reactant

### PR DESCRIPTION
**Pull request by avik-pal**
Much less intrusive than https://github.com/LuxDL/Lux.jl/pull/673. We automatically check for `get_device_type` and use Reactant if paired with Enzyme.

## TODOs

- [x] Reactant Backend -- not an actual backend but trigger with a combination of `xla_device` and `AutoEnzyme`
  - [x] `single_train_step!`
  - [x] `single_train_step`
  - [x] `compute_gradients`
- [x] Optimisers are not compiling. We have 2 solutions here
  - ~We hack around a proxy for the optimizer similar to what I was doing in #673.~
  - Actually stare at the error and fix it in Reactant proper (I kind of prefer going down this route) -- https://github.com/EnzymeAD/Reactant.jl/pull/156
- [x] Upstream
  - [x] https://github.com/EnzymeAD/Reactant.jl/pull/161
- [x] Testing
  - [x] Common model compile testing
  - [x] loss function compilation
  - [x] re-enable all tests

## Future PRs

- Docs will be added in #967 
- Inference? What would be a good API for that

### CODE REVIEW

**Comment by avik-pal** [comment]
## Example Usage

```julia
using Lux, Reactant, Enzyme, Random
using Optimisers

dev = xla_device()

model = Chain(
    Dense(2 => 32, tanh),
    Dense(32 => 32, tanh),
    Dense(32 => 2)
)
ps, st = Lux.setup(Random.default_rng(), model) |> dev

x = randn(Float32, 2, 1024) |> dev
y = x .^ 2 .+ 1

function sse(model, ps, st, (x, y))
    z, stₙ = model(x, ps, st)
    diff = z .- y
    return sum(abs2, diff), stₙ, (;)
end

train_state = Training.TrainState(model, ps, st, Descent(0.01))

grads, loss, stats, train_state_compiled = Training.single_train_step!(
    AutoEnzyme(),
    sse,
    (x, y),
    train_state
)

grads, loss, stats, train_state_compiled = Training.single_train_step!(
    AutoEnzyme(),
    sse,
    (x, y),
    train_state_compiled
)
```

Using any other ADType will cause an error.

**Comment by avik-pal** [comment]
https://github.com/EnzymeAD/Reactant.jl/pull/161 is needed for the loss functions tests to pass
